File: Day 13/RemoveKDigits.java
import java.util.LinkedList;

class Solution {
    public String removeKdigits(String num, int k) {
        LinkedList<Character> st = new LinkedList<>();
        StringBuffer ans = new StringBuffer();

        for(int i=0; i<num.length(); i++){
            while(st.size() != 0 && k != 0 && st.getFirst() > num.charAt(i)){
                st.removeFirst();
                k--;
            }
            if(st.size() != 0 || num.charAt(i) != '0'){
                st.addFirst(num.charAt(i));
            }
        }
        while(st.size() != 0 && k != 0){
            st.removeFirst();
            k--;
        }

        if(st.size() == 0){
            return "0";
        }

        while(st.size() != 0){
            ans.append(st.getFirst());
            st.removeFirst();
        }
        ans.reverse();

        return ans.toString();
    }
}

File: Day 13/RemoveKDigits.cpp
#include<bits/stdc++.h>
using namespace std;


class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        stack<char> st;
        for(int i=0; i<num.size(); i++){
            while(!st.empty() && k && st.top() > num[i]){
                st.pop();
                k--;
            }
            if(!st.empty() || num[i] != '0'){
                st.push(num[i]);
            }
        }
        while(!st.empty() && k){
            st.pop();
            k--;
        }

        if(st.empty()){
            return "0";
        }

        while(!st.empty()){
            num[n-1] = st.top();
            st.pop();
            n--;
        }

        return num.substr(n);
    }
};